packages/flutter/lib/src/material/floating_action_button_location.dart

```diff
@@ -584,7 +584,13 @@ mixin FabDockedOffsetY on StandardFabLocation {
     final double bottomSheetHeight = scaffoldGeometry.bottomSheetSize.height;
     final double fabHeight = scaffoldGeometry.floatingActionButtonSize.height;
     final double snackBarHeight = scaffoldGeometry.snackBarSize.height;
-    final double safeMargin = bottomViewPadding > contentMargin ? bottomViewPadding : 0.0;
+    final double bottomMinInset = scaffoldGeometry.minInsets.bottom;
+
+    final double safeMargin = bottomMinInset + fabHeight / 2.0 < contentMargin
+        ? 0.0
+        : bottomMinInset == 0.0
+            ? bottomViewPadding
+            : fabHeight / 2.0 + kFloatingActionButtonMargin;
```

Comment:
This is a bit hard to read and understand the decision path here. Can you break this out and add comments to explain what's going on?

Reply:
The code tries to apply the proper `safeMargin` according to the following criteria:

```txt
if contentMargin > bottomMinInset + fabHeight / 2   (Scaffold.bottomNavigationBar is bigger in height than
                                                 the keyboard, there is ample space to display a docked FAB)
  then safeMargin = 0.0
else if keyboard is closed
  then safeMargin = bottomViewPadding       (= 34.0px in iOS, 0.0 in Android)
else
  safeMargin = fabHeight / 2 + kFloatingActionButtonMargin  (this makes the docked FAB to behave like a 
                                                            floating FAB above the keyboard)
```

Additional context is at https://github.com/flutter/flutter/issues/26100#issuecomment-792810196

The usage of the conditional operator probably makes it a bit less readable. Would it be better to add comments or convert the code to an if-else statement here?

Reply:
Are the following comments sufficient?

```dart
    final double safeMargin =
      // Check whether bottomMinInset is lower than contentMargin, just enough
      // to display a docked FAB
      // bottomMinInset is typically represented by the software keyboard
      bottomMinInset + fabHeight / 2.0 < contentMargin
        // If low enough, don't provide a margin(don't move the docked FAB)
        ? 0.0
        // If not low enough, check whether bottomMinInset is zero(the software
        // keyboard is closed)
        : bottomMinInset == 0.0
            // If bottomMinInset is zero, provide bottomViewPadding as margin
            ? bottomViewPadding
            // If bottomMinInset isn't zero, provide a margin that would shift
            // the docked FAB enough so that it would stay away from the keyboard
            : fabHeight / 2.0 + kFloatingActionButtonMargin;
```

Reply:
> ```dart
>     double safeMargin;
> 
>     if (contentMargin > bottomMinInset + fabHeight / 2.0) {
>       // If contentMargin is higher than bottomMinInset enough to display a
>       // docked FAB without clipping, don't provide a margin
>       safeMargin = 0.0;
>     } else if (bottomMinInset == 0.0) {
>       // If bottomMinInset is zero(the software keyboard is not on the screen)
>       // provide bottomViewPadding as margin
>       safeMargin = bottomViewPadding;
>     } else {
>       // Provide a margin that would shift the docked FAB enough so that it
>       // stays away from the keyboard
>       safeMargin = fabHeight / 2.0 + kFloatingActionButtonMargin;
>     }
> ```

This LGTM, while the other is pretty trim and streamlined, which I appreciate, this is easiest to read and follow along with.